### ISSUE TITLE
config parameter

### ISSUE THREAD

**Issue by bcube2** [mfo13/mfo13.github.io#7]
Not really an issue, rather a question: could the config parameter be added?

**Comment by mfo13** [comment]
Yes. I think so. If it is just an space to place the name of the config file it is quite easy. On the other hand, perhaps, you are talking about the config tool generating a config file together an html code to call it? It is also possible but will take some time to implement. In this last case the user would have, lets say, a check box to generate the html code or the config file? By the way, just a curiosity, which one has precedence, the config file or the html specifications? Suppose you have bgcolor in both with different values, for example.

**Comment by bcube2** [comment]
Well, I know nothing about which approach will be easier to implement, but I was think like creating a text area similar to the initmove parameter and labelling it AnimCube.txt (or whatever .txt you like). If that textarea is not empty, the HTML code below the simulator would show "AnimCubeX("id=cube&config=AnimCube.txt&other_parameters") with loading the settings from that AnimCube.txt textarea (or a file which would be created?).

ACJS doc for the config parameter is saying "All values specified in the configuration file can be overwritten by simulator parameters declared in the HTML document (web page). That is, simulator parameters declared in the web page are prioritized over the parameters specified in the configuration file."

**Comment by mfo13** [comment]
OK. Understood! I'll try to implement your idea after fixing the other bugs. I think it is a very nice idea! The user would type the config parameters in a blank area and the config tool would produce both, the html code as well as the config.txt file text. So the user would just copy and paste both in his/her html file as well as the txt file. Great!

**Comment by mfo13** [comment]
I think we can start thinking in more detail about this tool. First of all, do you know what commands we can put in the config file? All of them? Even move, for example? Do I need to inspect the code first?

**Comment by mfo13** [comment]
Take a look at https://mfo13.github.io/index2.html
It is a first trial. What do you think?

**Comment by bcube2** [comment]
> First of all, do you know what commands we can put in the config file? All of them? Even move, for example? Do I need to inspect the code first?

Assuming that by commands you mean parameter settings, yes, the config file should support all of them. Nothing wrong with declaring move=R in the config file.

> Take a look at https://mfo13.github.io/index2.html
It is a first trial. What do you think?

Looks like the index.html file only with the Config file box added to me :-)

I would display that Config file box only if the config textarea is not empty. I would insert the config textarea (I would use the same design as for the initmove/initrevmove parameter) into the Cube Settings section just below the frame height option. I would rename "Config file" in the box below the simulator to "AnimCube.txt". The seetings in the AnimCube.txt box should be displayed [in this format](https://animcubejs.cubing.net/AnimCube3.txt). If a user clicks on the config parameter option in the Cube Settings section, "AnimCube.txt" label should appear somewhere. One way to do it is typing "AnimCube.txt configuration file, set parameters here." in gray color into that textarea, and when user cliks/taps to that textarea, this message will disappear.

**Comment by mfo13** [comment]
Yes! We are thinking pretty close the same. I've Just sent you to show it is possible and not so difficult to implement.

**Comment by mfo13** [comment]
Finished my idea. Take a look: https://mfo13.github.io/index2.html

**Comment by bcube2** [comment]
Thank you for trying, it's great to see the progress.

I have several suggestions for improving both functionality and design. Let's discuss one thing at a time, ok?

So should I start with functionality or desing suggestion?

**Comment by bcube2** [comment]
I'll go with this one: in the Cube Settings section, I suggest to hide the config text area by default, just like it is done in case of initmove.

And I'll add another one: I suggest to make that text area as big as in case of initmove, not just a single line.

**Comment by bcube2** [comment]
Actually, let me first check we are still on the same page and I am not asking for a functionality that can not be made.

Basically, I am asking for the config textarea to act like a .txt configuration file. That´s it, nothing else ;-)

If that´s possible, the AnimCube.txt tab below the simulator is not needed at all (and I will get to it later). I am also thinking about allowing users to upload the .txt configuration file (as an alternative to creating one via the config textarea), but I will get to it later as well.

**Comment by mfo13** [comment]
Well..., ok, we can make a text area that behaves like a config file, but you did not get my idea... let me explain. 
Why should the user produce by himself all the complex commands of the configuration file if he has a configuration tool?? The configuration tool can do that for him! It is what the config tool is doing right now. You have two possible outputs, one is the html code and the other one is the content of the AnimCube.txt file. That's why you have two output tabs now (notice the AnimCube.txt tab is not an input). The line in the cube setting, which is an input, is placed there for the user's config file name if he has one. As stated in the docs, the cofig parameter accepts only a file name, nothing else. OK, I understand your point, a space for the user pasting the parameters of his config file as an input to see what happens, ok,  it needs to be implemented, perhaps with an upload as you said instead of a blank area, I think it is better and less confusing. Did you get my point now?

**Comment by bcube2** [comment]
Oh, wow, thank you for your explanation. Now I get it.

It served as an IQ test and I failed, but with "hints" (placeholder in textarea) or additional/explanatory text it would be more clear what is expected from the user. It is working great.

Yeah, so your idea was to let the config tool create the .txt file for the user, while my idea was to input/paste declared parameters into the textarea and let the config tool process them. 

Let me think a bit about which approach is better.

**Comment by mfo13** [comment]
Why not both?

**Comment by bcube2** [comment]
I don´t see a way for your approach how to set the same parameter in both the configuration file and the HTML page (config tool in this case). Is it impossible or do I miss something?

example: in the configuration file there is buttonheight=15, in your config tool there is buttonheight=20

**Comment by mfo13** [comment]
OK. Notice that, when the user selects the AnimCube.txt tab the config input is disabled and turnned off. Now all settings of the tool is to produce a .txt file content. When the user selects the html code tab the config txt input is the one he provides as an input. The config tool now behaves like two tools in one. Actually I'm thinking in placing a blank aree for config input content, so the user can feedback with the information generated by the config tool itself.

**Comment by bcube2** [comment]
Let´s compare both approaches.

1. Your is introducing config textarea/input in the Cube Settings section, however, it is rather a configuration file name option than a configuration file option. Mine is introducing config textarea/input in the Cube Settings section which is an equivalent of the configuration file.
2. Your is introducing the AnimCube.txt tab below the simulator, mine is not introducing that tab (if clicked on config option in the Cube Settings section, a text like "Upload (upload being a link) AnimCube.txt configuration file or set parameters below" would appear - just like "Click on arrows to turn them. " for superfacelets. A multirow placeholder (note: this may not work in Safari) like 

```
parameter1=value1
parameter2=value2
parameter3=value3

(set as many parameters as you want)
```

would be typed in the textarea to hint the user a right input format)

  3. Upload is not user friendly for mobile phone users, that´s why there is the textarea to input settings.
  4. Your makes it impossible to set the same parameter with two different values, mine makes it possible

> Why should the user produce by himself all the complex commands of the configuration file if he has a configuration tool?? The configuration tool can do that for him!

Exactly, all the settings can be copied from the HTML code tab already for my approach (granted, the user would have to change its format). 

So right now I am more inclined to my approach, however, I have no idea how difficult it would be to implement in comparison with your approach.

**Comment by mfo13** [comment]
Welll... answers bellow:
1. No, I intend the same as you, not only the config file name but also its content as an input.
2. My approach is both.
3. Agree, text area is better.
4. Mine too, because it is your input approach side by side with my output approach, it is going to be two tools in one, parameters will not mix. I'll work on that and show, I think it is better than eplaining with words, if you don't like it, so far so good, we just erase it.

**Comment by bcube2** [comment]
I'll wait and see what you can come up with then :-)

Don't delete anything yet, I was just expressing my opinion so far. But as I see, you are still working on it.

**Comment by mfo13** [comment]
Done, but... I'm guessing what you are going to say... and if you say that, well..., the implementation is going to be hard. Since may vacations are going to an end you'll need to make a decision about what to keep for this version and what to leave for the next year.

**Comment by bcube2** [comment]
Looking good, thank you for your effort. I made a slight change at https://bcube2.github.io/test6/sources/codes/enhancement/parameters/index2.html, and now I see only 2 problems:

1. line breaks in placeholder won´t work in Safari according to https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#browser_compatibility (okay, I can live with that)
2. I still don´t see the benefits of having AnimCube.txt tab, so I will see if I can get rid of it by combinanig what´s in your index.html file and the config functionality in your index2.html file

> you'll need to make a decision about what to keep for this version and what to leave for the next year.

As I was saying, no timelines are given :-)

I plan to update ACJS doc with this version soon: https://bcube2.github.io/test6/sources/codes/enhancement/parameters/config_tool.html

**Comment by mfo13** [comment]
OK. When you make a decision just tell me. When I have time I'll keep working in my original idea, two tools in one.

**Comment by bcube2** [comment]
> now I see only 2 problems

Another problem is that once a setting like supercube=1 is declared in the config textarea, it can not be undone in the config tool, even if it says supercube off. Possible solution: set all parameters in the config tool to default values using checkboxes. The question is whether it is worth it in terms of both programming difficulty and user friendliness.

> line breaks in placeholder won´t work in Safari

This could work even for Safari users: https://bcube2.github.io/test6/sources/codes/enhancement/parameters/index2.html

I have updated https://animcubejs.cubing.net/sources/codes/enhancement/parameters/config_tool.html. For me, my biggest problem is with the AnimCube.txt tab in your index2.html file. It feels really unintuitive - if I enter parameters into the config textarea, I would expect they will be shown in the AnimCube.txt tab. Why this tab contains the settings from the config tool is not clear to me (as a newcomer) either. Maybe renaming that tab to something more meaningful could help. Or maybe a link to that tab from the home page could help (while not displaying that tab on the home page). But right now I feel it is not as simple as it could be for newcomers (note that this is my personal opinion, I could be totally wrong). Outputting the right format for the config file is a nice feature, but displaying a new tab only because of that is more confusing than simplifying to me. I am terribly sorry for bad news :-( (but don´t take it personally, it is only my opinion) 

Combining that with a supercube=1-like settings mentioned above, I decided to not update the config_tool.html with the config feature yet. Maybe in the future :-)

In any case, thank you very much for a nice tool. Its functionality should be matched with the other tool (https://animcubejs.cubing.net/sources/codes/enhancement/parameters/create.html) now.

**Comment by mfo13** [comment]
OK. We need to think better about the introduction of config, input and output. One possibility for the input, as you said, is to place default boxes everywhere... but I don't like it, it looks ugly for me...

**Comment by bcube2** [comment]
> One possibility for the input, as you said, is to place default boxes everywhere... but I don't like it, it looks ugly for me...

I can think of another way but it is way too complicated to me to even think about, not to mention to implement it programatically... The idea would be to somehow set the value from the config textarea to the config tool, so that the value could be changed back in the config tool. If you think about it in detail, it gets pretty complicated quickly (at least for me).

**Comment by mfo13** [comment]
Well, it is complicated for me but doable. I can implement a kind of parsing back function. But it doesn't solve our problem because the dafault for supercube is already 0. The point is this, with a config saying that supercube=1 this means that the default has changed for this user, do you understande? So I think, more than parsing back the parameters, the defaults need to be changed... even all of them deppending on the config file...

**Comment by bcube2** [comment]
> But it doesn't solve our problem because the dafault for supercube is already 0. The point is this, with a config saying that supercube=1 this means that the default has changed for this user, do you understande?

Actually, I don´t :-) If a user sets supercube=1 in the config file, he would see supercube on in the config tool, but the output written below the simulator would show only config=AnimCube.txt (that is, without the supercube=1 setting). 

If a user wants to change back to normal cube, he would set supercube off in the config tool, and the output would be config=AnimCube.txt&supercube=0.

This approach is breaking the rules we have now when the config parameter is not being used, hence I wrote it gets complicated soon.

**Comment by mfo13** [comment]
> Yes, exactly. Internally the script assumes  0 is the default, so now we need a default variable that changes with config.

**Comment by bcube2** [comment]
If you put it that way, it is a lot easier to visualize the whole process for me :-)

**Comment by bcube2** [comment]
I have been revisiting the ideas discussed above and I think we need to take one step back in order to keep forward here.

I would like to propose a two phase process:

Phase 1 (critical) - getting the functionality working
Phase 2 (optional) - deciding if and how to introduce a way of the tool interacting with the configuration file

Phase 2 will not be considered further in this comment, we can talk about it later once the phase 1 is finished. As for phase 1, the functionality can be tested against https://animcubejs.cubing.net/sources/codes/enhancement/parameters/create.html with config=../../../../AnimCube3.txt&bgcolor=124578

I can think of 2 ways how to get the functionality working:

1) let the user create and name the configuration file, save it to his device, then upload it from his device (see https://mfeather1.github.io/3ColorCube/solver_qc/save_file.html which is working fine on my cell phone)

2) don´t work with the physical configuration file at all, convert the "server" syntax from the configuration file to the "client" syntax in the tool as described at https://animcubejs.cubing.net/animcubejs.html#config (see the "Show how to do it" dropdown).

In phase 1 at https://mfo13.github.io/index2.html, the AnimCube.txt tab below the simulator doesn´t need to be there at all, and the AnimCube.txt text in the config option doesn´t need to be there as well. 

Assuming option 2 (no working with a physical configuration file required) is the way to go from a programming side, if the config textarea is not empty, I would put "config=configuration_file.txt" instead of the current "config=AnimCube.txt" to the HTML code below the simulator,

The last thing I would like to address in phase 1 is disabling the settings written in the config textarea (for option 2 above). If the textearea is not empty, a checkbox "ignore configuration file" could appear, and checking it would result in making the textearea read-only with faded-out text color and disappearing of the "config=confuguration_file.txt" part from the HTML code text below the simulator. 

---------------------
Edit: wait a second, I didn´t realize phase 1 is basically done (AnimCube.txt tab is used instead of checkbox but the functionality of those two features is basically the same) so never mind this comment.

**Comment by mfo13** [comment]
Mind connection??!!! I was exactly revisiting these old ideas yesterday! :)) Moreover, I was thinking in placing the config_tool.html in a more modern frame configuration, using bootstrap, so it could be easy to access in the desktop as well as the mobile. I think it would be another phase, 3 or 0? Phase 1 is implemented but not working properly, remember? The default values should change with the input of the txt file.

**Comment by bcube2** [comment]
> The default values should change with the input of the txt file.

Oh, right. I can see that if I set bgcolor=124578 in the config textarea I can´t set it back to default using the checkbox in the tool. Bummer.

> Mind connection??!!! I was exactly revisiting these old ideas yesterday! :))

As they say, great minds think alike :-)

**Comment by mfo13** [comment]
Dear bcube2, would like to test the responsive layout for cell phones or smaller screens? [https://mfo13.github.io/index3.html](https://mfo13.github.io/index3.html)

**Comment by bcube2** [comment]
Regarding graphical design/interface/layout, I am afraid I can not be helpful much.

One difference in graphical design I noticed is that in Firefox on desktop Windows, there are violet horizontal lines separating the groups of settings (Cube, Cube Settings, Buttonbar, Hint, Move, Faces), while in the other browsers I tried (Chrome on desktop Windows, mobile Chrome on Android, Samsung Internet on Android) these lines are not violet. Not sure if this is intended.

**Comment by mfo13** [comment]
Dear bcube2,

   Oh! The violet line was a test I forgot to clean... done! Design and fancy things are not my business as well, I'm more concerned if it works better now in cell phones and small screens. Is that ok for you? Thank you!

**Comment by bcube2** [comment]
> I'm more concerned if it works better now in cell phones and small screens. Is that ok for you?

As I see it, the ACJS configuration tool at https://animcubejs.cubing.net/sources/codes/enhancement/parameters/config_tool.html is supposed to be visited mainly by newcomers/ACJS beginners or by users who want to experiment quickly with different settings.

In this target group, I would imagine the users interact with both the settings on the left and/or the simulator on the right. The hamburger menu as presented at https://mfo13.github.io/index3.html prevents users from this interactivity feature (remember, the users are beginners, so I would expect they would like to experiment with the settings a lot) because while 1 or 2 clicks/taps on the hamburger symbol is fine, the new version forces them to click/tap on the hamburger symbol each time they want to experiment with a different setting.

In the https://animcubejs.cubing.net/sources/codes/enhancement/parameters/config_tool.html version this can be addressed by zooming in/out and/or putting the phone to the landscape mode (from the portrait mode) if needed/wanted.

Because of this, I personally don´t like the new version better than the old version. However, you are in charge with upgrading the ACJS configuration tool so I won´t stop you with upgrading the tool if you think the new version is better for the users. There are a few functionality issues in the new version I noticed - the most important being the config parameter not fully functional - so the new version is not ready for a production yet anyway in my opinion.

**Comment by mfo13** [comment]
Thank you! I was exactly in doubt about that, what would be better for a beginner? Zooming in out in the cell phone or clicking the menu link all the time? So, let's keep as it is and we have another option in the future, just in case.

**Comment by mfo13** [comment]
Dear @bcube2, finally I have a version for your tests regarding the input of config parameter (config txt file). I'd have many things to say about this version but I think it is better just to let you go through the trials, find bugs, report preferences and so on... (just acces the main webpage of this repository). Regards!

**Comment by bcube2** [comment]
> just acces the main webpage of this repository

Just to be sure we are on the same page, I have been visiting https://mfo13.github.io/

When I typed "snap=1" to the config file text area, I got "AnimCube3("config=AnimCube.txt&snap=0")" in the HTML code. The "&snap=0" part should not be there. 

Similarly, when I typed "scale=5" to the config file text area, I got "AnimCube3("config=AnimCube.txt&scale=0")" in the HTML code. The "&scale=0" part should not be there.

Lastly I typed "perspective=200" to the config file text area. I got "AnimCube3("config=AnimCube.txt&perspective=2")" in the HTML code. The "&perspective=2" part should not be there.

I am sure you can see the pattern. It seems I am missing your logic (I have no idea why I don´t get only "AnimCube3("config=AnimCube.txt")" in the HTML code and you are missing my logic :-)

Why those settings should not be there: because when they are in there, the simulator is not reflecting the settings written in the config file.

**Comment by mfo13** [comment]
Dear @bcube2, well, all of these is a question of interpretation. Note that these parameters don't have the default option, so what is the priority? The config or the command line? According to the Animcube documentation the command line has priority, therefore, if you set something diferent of the usual default value in the config (a new default) the selection in the parameter will not be changed because it has the priority, the appearance of the cube will not change as well, and thus the command line in the html output changes accordingly to this priority, a parameter different from the default. Notice that this does not happen when you have a default parameter set. Try, for example, unset the default of colorscheme and change the default value in the config file. The cube appearence will not change at all but the html code will now reflect your command line option wich is 'wyorb' because it is not the default value anymore and has the priority. However, if you set a new colorscheme in config but with the default colorscheme set the appearence of the cube will promptly change and you'll see, in the command line, just the explicit config parameter without any colorschme, because you are saying you want the default. I understand it appers a bit confusing, perhaps we should give the option for setting defaults for every single parameter...??

**Comment by bcube2** [comment]
It indeed seems we are interpreting some things differently.

> Note that these parameters don't have the default option

Assuming that by option you mean value, the snap and scale parameters have a default value 0, and perspective parameter has a default value 2. 

> so what is the priority? The config or the command line?

ACJS doc says: "All values specified in the configuration file can be overwritten by simulator parameters declared in the HTML document (web page). That is, simulator parameters declared in the web page are prioritized over the parameters specified in the configuration file." What it means is that if an undeclared parameter in the web page is declared in the config file, the simulator will use the setting set in the config file. But if the same parameter is declared in both the web page and the config file, the simulator will use the setting set in the web page. So my goal is to match this logic (which you can experience as a webmaster creating a simulator in a web page) in your configuration tool.

> According to the Animcube documentation the command line has priority, therefore, if you set something diferent of the usual default value in the config (a new default) the selection in the parameter will not be changed because it has the priority

Please try to replicate what you said as a webmaster. Declare only the config parameter in your web page, and write this line to your config file: snap=1. You are saying a half-turned layer should stay where it is, I am saying it should snap to the corner.

Let´s see if we can agree on this first, we can disscuss the rest of your post later.

**Comment by mfo13** [comment]
OK. I've been thinking about... and my opinion is just to place a default checkbox in every parameter. Some of them don't need this, actually, like move or initmove.

**Comment by bcube2** [comment]
Maybe that's an overkill. I mean, the config parameter is supposed to be used for several cubes (like CSS file is used for several HTML pages) but Config Tool can display only 1 cube anyway.

What problem would be solved by doing that? Remember, a parameter set to default value is not written in the HTML code below the simulator.

If a webmaster wants to use a default value of some parameter, he will not probably set it to the default value - instead he will just use its default value without setting it.

I would say KIS (keep it simple): if a user of Config Tool wants to use snap=0 when he is declaring snap=1 in his config file, he can just delete the snap=1 setting in his config file instead of clicking on the default checkbox which would add "&snap=0" (i.e. its default setting) to the HTML code bellow the simulator. 

Not to mention, options "on", "off" and "default" for the snap parameter would be confusing - ACJS doc says there are 2 possible values, but Config Tool would offer 3. It's not obvious at all what's the difference between "off" and "default" for users not thinking about all possibilities with the config parameter.

I can see where you are going with the default checkbox for every parameter, but introducing it may introduce other issues as highlighted above, hence my KIS.

**Comment by mfo13** [comment]
OK. So the current behavior of snap is correct. The original default is snap=0 (off). If the user change it in the menu (on) then the html will change promptly with snap=1. Now, if the user sets his config file with snap=1, automaticaly the html will not display snap=1, since now it is the new default value and it is selected in the menu. Ohtherwise, if the user sets the config with snap=0 then the html will display snap=1 because the new default is 0. I'm assuming the options set in the menu is what the user wants to see in the cube, so they have the priority over config. The menu is the html command line, it has the priority over the default values, it doesn't matter if the default is the original one or another defined in the config.

**Comment by bcube2** [comment]
Below is what I am getting and what I am expecting for your scenario.

> The original default is snap=0 (off). If the user change it in the menu (on) then the html will change promptly with snap=1.

case: snap set to on
I am getting: AnimCube3("snap=1")
I am expecting: AnimCube3("snap=1")

> Now, if the user sets his config file with snap=1, automaticaly the html will not display snap=1, since now it is the new default value and it is selected in the menu.

case: snap set to on and snap=1 in config file
I am getting: AnimCube3("config=AnimCube.txt")
I am expecting: AnimCube3("config=AnimCube.txt&snap=1")

Reasoning: the snap parameter is declared in both the menu and the config file. In such case, menu declaration is prioritized, therefore the simulator will use the snap=1 setting from the menu.

> Ohtherwise, if the user sets the config with snap=0 then the html will display snap=1 because the new default is 0.

case: snap set to on and snap=0 in config file
I am getting: AnimCube3("config=AnimCube.txt&snap=1")
I am expecting: AnimCube3("config=AnimCube.txt&snap=1")

Reasoning: the snap parameter is declared in both the menu and the config file. In such case, menu declaration is prioritized, therefore the simulator will use the snap=1 setting from the menu.

---

case: snap set to off and snap=1 in config file
I am getting: AnimCube3("config=AnimCube.txt&snap=0")
I am expecting: AnimCube3("config=AnimCube.txt")

Reasoning: we don´t know whether the Config Tool user wishes to use the default snap parameter value from the menu or whether he explicitly wants to declare snap=0 (as a webmaster can do in his own HTML page). Since menu parameters set to default values are treated as no settings at all in Config Tool, the snap set to off option is expected to be treated as no setting at all, therefore the case can be simplified to only snap=1 in config file.


P.S. if you have your own logic that is not compatible with my logic, that´s perfectly fine. I am not making my own decisions, I am simply following the rules given by the program. So if you want to make the rules which are more logical to you, there would be no reason to continue any further as the problem would be unsolvable.

bonus case: snap set to off and snap=0 in config file:
I am getting: AnimCube3("config=AnimCube.txt")
I am expecting: AnimCube3("config=AnimCube.txt")

Reasoning: we don´t know whether the Config Tool user wishes to use the default snap parameter value from the menu or whether he explicitly wants to declare snap=0 (as a webmaster can do in his own HTML page). Since menu parameters set to default values are treated as no settings at all in Config Tool, the snap set to off option is expected to be treated as no setting at all. Therefore the simulator will use the snap=0 setting from config file.

**Comment by mfo13** [comment]
So, you mean that everything that is the config should not be taken as default? If this is the case the phrase in config tool "AnimCube.txt (override default settings)" has no meaning. I think that's the difference between our logic. I'm assuming that the user wants to change the default values when setting config. It makes sense to me. Imagine you are building a website with many cubes with a default setup different from the original, you just adjust one configfile and you are done. If some few parameters in one page or another needs a different setting then you use the html command line. So, it is ok to set snap=1 in the config file and you don't need to care about that in the whole website because it is now your current default; when needed you just use snap=0 in the html command line.

**Comment by bcube2** [comment]
> So, you mean that everything that is the config should not be taken as default?

Nope. The decision tree is following. Is a menu parameter set to a value different than default? If so, the simulator will use that value no matter what, you don't need to check the config file for that specific parameter at all. If not, meaning a menu parameter is set to its default value, then if that parameter is declared in the config file, the simulator will use the declaration of that parameter from the config file.

One other difference in our logics seems to be that you are thinking in sequences when declaring parameters, while I am thinking in terms of whole 1-time declarations. You earlier said "set snap to on. Now, if the user sets his config file with snap=1, automaticaly the html will not display snap=1, since now it is the new default value and it is selected in the menu." A webmaster works with the simulator with declared parameters, so he will skip this sequential process of setting one parameter at a time then setting the other parameter at a time. Instead, he works with the settings all set at the same time: config=AnimCube.txt&snap=1

I agree with the rest of your post :-)

**Comment by mfo13** [comment]
"Nope. The decision tree is following. Is a menu parameter set to a value different than default?"

    Which default? The user can set new defaults with config.

    You say there is just one default, the one from AnimCube and config as well as html line are just command lines that can, eventualy, replace each other. I don't think so, because html has allways precedence over config however config has allways the precedence to the original defaults.

**Comment by mfo13** [comment]
I was thinking about your point of view. We can do like that, but, in this case, we should change the default boxes to another denomination, like open/close or so. Otherwise it will be confusing. Take this example, the user sets config with colorscheme=01234 and the default colorscheme box is already checked. The cube will change appearence because of config and the default box is already checked, that looks incoherent. Otherwise, with the default box checked the cube do not change but, in the html line it appears colorscheme=wyorgb in front of config, ok, that sounds correct.

**Comment by bcube2** [comment]
> Which default?

The one stated in the ACJS doc.

> You say there is just one default, the one from AnimCube and config as well as html line are just command lines that can, eventualy, replace each other.

I am sorry, I don´t understand this part. Feel free to rephrase it if you want.

> html has allways precedence over config however config has allways the precedence to the original defaults.

Agreed.

---

It seems it all boiles down to a different perception of what is a default value. I understand it as a value described in the ACJS doc, and there is only one such value for (almost) all parameters. In other words, a parameter can not have 2 or more default values according to the doc. So when declaring a parameter in the config file, I am not thinking of resetting it to another default value (the doc implies there is only 1 such value), instead I am thinking of it as setting a new value (because 2 or more values can be set to a parameter).

We mean the same thing, only using different vocabulary.

**Comment by mfo13** [comment]
OK. So let's keep just one immutable default value. No problem. How are we going to proceed with the default checkboxes? We keep them and when the user sets colorscheme=01234 in the config nothing will change on the cube but, rather, the colorscheme=wyorgb appears in the html? Or, we just avoid using this therm "default" and we change it to another thing like "open", "open details"....?

**Comment by bcube2** [comment]
> Take this example, the user sets config with colorscheme=01234 and the default colorscheme box is already checked. The cube will change appearence because of config and the default box is already checked, that looks incoherent.

What exactly looks incoherent? You start from default value, which is wyorgb. You set it to 012345 via config file and see its changing effect on the simulator. Looks pretty coherent to me :-)

**Comment by mfo13** [comment]
No, because in the menu the default box is selected but the default value is not displayed on the cube. Do you think it makes sense having a default box set by the user in the menu with a completely different scheme on the cube without any html command for that? It sounds weired to me.

**Comment by mfo13** [comment]
Remember, I'm assuming that what the user selects in the menu is what he wants. In your point of view, he wants the default but is just overriding it with config??

**Comment by bcube2** [comment]
> How are we going to proceed with the default checkboxes? We keep them and when the user sets colorscheme=01234 in the config nothing will change on the cube but, rather, the colorscheme=wyorgb appears in the html?

case: colorscheme set to default in config tool and colorscheme=012345 is the config file
I am expecting: AnimCube3("config=AnimCube.txt")

Default settings (like colorscheme=wyorgb) will never be shown in the HTML code below the simulator in config tool.

> No, because in the menu the default box is selected but the default value is not displayed on the cube. Do you think it makes sense having a default box set by the user in the menu with a completely different scheme on the cube without any html command for that?

As long as the user set it (by it I mean the colorscheme parameter) to a different value via config file, yup, totally. The html command, as you call it, will be config=AnimCube.txt.

> I'm assuming that what the user selects in the menu is what he wants.

But as I shown earlier, you will fail: https://github.com/mfo13/mfo13.github.io/issues/7#issuecomment-5077996879 (see the last reasoning, there are 2 possibilities, you can´t predict what the user wants).

> In your point of view, he wants the default but is just overriding it with config??

I am sorry, I don´t get this question. Please rephrase it if you want me to answer it.

**Comment by mfo13** [comment]
OK. I totally understood your point of view and the answer to my last question is yes because you said before: "Reasoning: we don´t know whether the Config Tool user wishes to use the default snap parameter value from the menu or whether he explicitly wants to declare snap=0 (as a webmaster can do in his own HTML page)." I think I've completely understood what to do, however it will take some time because I need to erase that programing in which config behaves like a new default.

**Comment by bcube2** [comment]
And I, in return, think I understand now what you meant by open/close checkboxes. Those are the 2 possibilities I was talking about earlier in the Reasoning paragraph.

Again - we mean the same thing (which is good), only using different vocabulary.

As for implementing, there is no time frame :-). Take your time, and if in doubt, feel free to consult. Neither of us wants your programming to be redone :-(.

**Comment by mfo13** [comment]
OK. Check the snap case now. I think it is working following your logic. If so, I can continue on the other parameters.

**Comment by bcube2** [comment]
All 4 snap cases are working perfectly now, good job. 👍

Even accidental misuse cases where the user sets the parameter more times with different values like

snap=0
snap=1

in the config file are working perfectly, well done. 👍

**Comment by mfo13** [comment]
Nice! I'll keep fixing. I think I'll let you know and check at each one, because some are a bit different, they have numbers, for example.

**Comment by bcube2** [comment]
Just an FYI: you made me realize to make this change: https://github.com/cubing/AnimCubeJS/commit/280f89a7f0e8c6719aae0aaecc4f1dc153892456

No change in your code is needed.
